File: answers/leetcode/Add and Search Word - Data structure design/Add and Search Word - Data structure design.java
class WordDictionary {
    class Trie {
    
        private TrieNode root;

        /** Initialize your data structure here. */
        public Trie() {
            this.root = new TrieNode();
        }

        /** Inserts a word into the trie. */
        public void insert(String word) {
            TrieNode current = this.root;
            for (char c : word.toCharArray()) {
                if (current.getChild(c) == null) {
                    TrieNode next = new TrieNode();
                    current.setChild(c, next);
                }
                current = current.getChild(c);
            }
            current.setEnd(true);
        }

        /** Returns if the word is in the trie. */
        public boolean search(String word) {
            return this.search(word, 0, this.root);
        }
        
        private boolean search(String word, int index, TrieNode current) {
            if (index >=  word.length()) {
                return current.isEnd();
            }
            char c = word.charAt(index);
            if (c == '.') {
                for (char i = 'a'; i < 'z'; i++) {
                    TrieNode next = current.getChild(i);
                    if (next != null && this.search(word, index + 1, next)) {
                        return true;
                    }
                }
                return false;
            } else {
                TrieNode next = current.getChild(c);
                return next != null && this.search(word, index + 1, next);
            }
        }

        /** Returns if there is any word in the trie that starts with the given prefix. */
        public boolean startsWith(String prefix) {
            TrieNode current = this.root;
            for (char c : prefix.toCharArray()) {
                TrieNode next = current.getChild(c);
                if (next == null) {
                    return false;
                } else {
                    current = next;
                }
            }
            return true;
        }

        private class TrieNode {

            private TrieNode[] children;
            private boolean isEnd;

            public TrieNode() {
                this.children = new TrieNode[26];
            }

            public void setChild(char c, TrieNode node) {
                this.children[c - 'a'] = node;
            }

            public TrieNode getChild(char c) {
                return this.children[c - 'a'];
            }

            public void setEnd(boolean isEnd) {
                this.isEnd = isEnd;
            }

            public boolean isEnd() {
                return this.isEnd;
            }
        }
    }

    /**
     * Your Trie object will be instantiated and called as such:
     * Trie obj = new Trie();
     * obj.insert(word);
     * boolean param_2 = obj.search(word);
     * boolean param_3 = obj.startsWith(prefix);
     */
    
    private Trie root;

    /** Initialize your data structure here. */
    public WordDictionary() {
        this.root = new Trie();
    }
    
    /** Adds a word into the data structure. */
    public void addWord(String word) {
        this.root.insert(word);
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(String word) {
        return this.root.search(word);
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: answers/leetcode/Add and Search Word - Data structure design/Add and Search Word - Data structure design.cpp
//@type Backtracking Trie Design
//@result 13 / 13 test cases passed. Status: Accepted Runtime: 160 ms Submitted: 0 minutes ago You are here! Your runtime beats 25.21% of cpp submissions.

#include <iostream>
#include <string>
using namespace std;

class TrieNode {
public:
    // Initialize your data structure here.
    TrieNode() : key_flag(false) {
        for (int i = 0; i < 26; ++ i) {
			children[i] = NULL;
		}
    }
	bool key_flag;
	TrieNode *children[26];
};

class WordDictionary {
public:
	WordDictionary(void) {
        root = new TrieNode();
	}
    // Adds a word into the data structure.
    void addWord(string word) {
        TrieNode *node = root;
		for (size_t i = 0; i < word.length(); ++ i) {
			if (! node->children[word[i] - 'a']) {
				node->children[word[i] - 'a'] = new TrieNode();
			}
			node = node->children[word[i] - 'a'];
		}
		node->key_flag = true;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    bool search(string word) {
		return search(word, root, 0);
    }
    bool search(string word, TrieNode *node, size_t pos) {
		if (pos >= word.length()) {
			return node ? node->key_flag : false;
		}
		else if (node) {
			if ('.' == word[pos]) {
				for (size_t i = 0; i < 26; ++ i) {
					if (search(word, node->children[i], pos + 1)) {
						return true;
					}
				}
				return false;
			}
			else {
				return search(word, node->children[word[pos] - 'a'], pos + 1);
			}
		}
		else {
			return false;
		}
	}
	TrieNode *root;
};

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

int main() {
	// your code goes here
	WordDictionary wordDictionary;
	wordDictionary.addWord("a");
	wordDictionary.addWord("ab");
	cout << wordDictionary.search("a") << endl;
	cout << wordDictionary.search("a.") << endl;
	cout << wordDictionary.search("ab") << endl;
	cout << wordDictionary.search(".a") << endl;
	cout << wordDictionary.search(".b") << endl;
	cout << wordDictionary.search("ab.") << endl;
	cout << wordDictionary.search(".") << endl;
	cout << wordDictionary.search("..") << endl;
	return 0;
}